### PR TITLE
Add `calloc`

### PR DESCRIPTION
**Pull request by TimDiekmann**
Closes rust-lang#679

### CODE REVIEW

**Comment by TimDiekmann** [comment]
I added `Calloc`, but since I haven't contributed to miri yet, I need some mentoring on the tests.

**Comment by RalfJung** [comment]
For a test, all you have to do is write a little program (with `fn main` and all) that calls `calloc`, and plays with it a bit to see if it is well-behaved -- like, see that it is really all 0. Then put that into a file in `tests/run-pass`.

File: src/fn_call.rs [hunk 1]
```diff
@@ -93,6 +93,30 @@ pub trait EvalContextExt<'a, 'mir, 'tcx: 'a + 'mir>: crate::MiriEvalContextExt<'
                     this.write_scalar(Scalar::Ptr(ptr.with_default_tag()), dest)?;
                 }
             }
+            "calloc" => {
+                let items = this.read_scalar(args[0])?.to_usize(this)?;
+                let count = this.read_scalar(args[1])?.to_usize(this)?;
+                let size = if let Some(size) = items.checked_add(count) {
```
Comment (RalfJung) [inline src/fn_call.rs]: Shouldn't this be `checked_mul`?

File: src/fn_call.rs [hunk 2]
```diff
@@ -93,6 +93,30 @@ pub trait EvalContextExt<'a, 'mir, 'tcx: 'a + 'mir>: crate::MiriEvalContextExt<'
...
+                    size
+                } else {
+                    return err!(MachineError(format!(
+                        "calloc: overflow of items * count: {} * {}",
+                        items, count,
+                    )));
```
Comment (RalfJung) [inline src/fn_call.rs]: You can use [`ok_or_else`](https://doc.rust-lang.org/nightly/std/option/enum.Option.html#method.ok_or_else) to make this more Rust-y.

File: src/fn_call.rs [hunk 3]
```diff
@@ -93,6 +93,30 @@ pub trait EvalContextExt<'a, 'mir, 'tcx: 'a + 'mir>: crate::MiriEvalContextExt<'
...
+                };
+                if size == 0 {
+                    this.write_null(dest)?;
+                } else {
+                    let align = this.tcx.data_layout.pointer_align.abi;
+                    let ptr = this.memory_mut()
+                        .allocate(Size::from_bytes(size), align, MiriMemoryKind::C.into())
```
Comment (RalfJung) [inline src/fn_call.rs]: The subexpression `Size::from_bytes(size)` appears twice, please let-bind it.

File: src/fn_call.rs (hunk 1 above)
Comment (TimDiekmann) [reply src/fn_call.rs]: Yes, thanks!

**Comment by TimDiekmann** [comment]
> For a test, all you have to do is write a little program (with `fn main` and all) that calls `calloc`, and plays with it a bit to see if it is well-behaved -- like, see that it is really all 0. Then put that into a file in `tests/run-pass`.

Ah ok, didn't think that would be enough.

File: src/fn_call.rs [hunk 4]
```diff
@@ -93,6 +93,21 @@ pub trait EvalContextExt<'a, 'mir, 'tcx: 'a + 'mir>: crate::MiriEvalContextExt<'
                     this.write_scalar(Scalar::Ptr(ptr.with_default_tag()), dest)?;
                 }
             }
+            "calloc" => {
+                let items = this.read_scalar(args[0])?.to_usize(this)?;
+                let len = this.read_scalar(args[1])?.to_usize(this)?;
+                let bytes = items.checked_mul(len).ok_or_else(|| InterpError::Overflow(mir::BinOp::Mul))?;
+
+                if bytes== 0 {
```
Comment (bjorn3) [inline src/fn_call.rs]:
```suggestion
                if bytes == 0 {
```

**Comment by TimDiekmann** [comment]
How long does it take to be provided by rustup? Next nightly?

**Comment by RalfJung** [comment]
The submodule in rustc needs to be updated, and then a nightly need to be released.

**Comment by TimDiekmann** [comment]
Is there some automation involved or does it need a PR?

**Comment by RalfJung** [comment]
There is no automation; it needs a PR in the rustc repo.

**Comment by TimDiekmann** [comment]
Okay, will do this! :+1:

**Comment by RalfJung** [comment]
Awesome, and thanks a lot :)

**Pull request merged**